core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java

```diff
@@ -0,0 +1,241 @@
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.util.internal.StringUtil;
+
+final class QueryStringEncoder {
+
+    // Forked from netty-4.1.43.
+    // https://github.com/netty/netty/blob/bd8cea644a07890f5bada18ddff0a849b58cd861/codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java
+
+    private static final byte WRITE_UTF_UNKNOWN = (byte) '?';
+    private static final char[] CHAR_MAP = "0123456789ABCDEF".toCharArray();
+    private static final boolean[] SAFE_OCTETS =
+            createSafeOctets("-_.*abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789");
+
+    private static boolean[] createSafeOctets(String safeChars) {
+        int maxChar = -1;
+        for (int i = 0; i < safeChars.length(); i++) {
+            maxChar = Math.max(safeChars.charAt(i), maxChar);
+        }
+        final boolean[] octets = new boolean[maxChar + 1];
+        for (int i = 0; i < safeChars.length(); i++) {
+            octets[safeChars.charAt(i)] = true;
+        }
+        return octets;
+    }
+
+    static void encodeParams(TemporaryThreadLocals tempThreadLocals,
+                             StringBuilder buf, QueryParamGetters params) {
+        for (Entry<String, String> e : params) {
+            final String name = e.getKey();
+            if (isSafeOctetsOnly(name)) {
+                buf.append(name);
+            } else {
+                encodeUtf8Component(tempThreadLocals, buf, name);
+            }
+            buf.append('=');
+
+            final String value = e.getValue();
+            if (isSafeOctetsOnly(value)) {
+                buf.append(value);
+            } else {
+                encodeUtf8Component(tempThreadLocals, buf, value);
+            }
+            buf.append('&');
+        }
+
+        buf.setLength(buf.length() - 1);
+    }
+
+    private static boolean isSafeOctetsOnly(String s) {
+        final int len = s.length();
+        for (int i = 0; i < len; i++) {
+            if (!isSafeOctet(s.charAt(i))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static boolean isSafeOctet(char c) {
+        return c < SAFE_OCTETS.length && SAFE_OCTETS[c];
+    }
+
+    /**
+     * Encodes a query component (name or value).
+     *
+     * @see ByteBufUtil#writeUtf8(ByteBuf, CharSequence, int, int)
+     */
+    private static void encodeUtf8Component(TemporaryThreadLocals tempThreadLocals,
+                                            StringBuilder buf, String s) {
+        final char[] tmp = tempThreadLocals.charArray(12);
+        int safeOctetStart = 0;
+        final int len = s.length();
+        for (int i = 0; i < len; i++) {
+            final char c = s.charAt(i);
+            if (c < 0x80) {
+                if (!isSafeOctet(c)) {
+                    if (i > safeOctetStart) {
+                        buf.append(s, safeOctetStart, i);
```

Comment:
This `safeOctetStart` was just an idea, but I don't know if it helps or hurts in practice given the extra branches 😅 It is very much dependent on the content, and while I think for longer strings (which we'd expect from forms) it could help, for now, probably better to keep it simpler for easier comparison to other implementations, and we can revisit later. It's probably worth doing if it doesn't negatively impact numbers too much as I'd expect it to have small negative impact on small input, large positive impact on larger input.

Reply:
I don't know. At the moment we lose at decoding long query string although the gap is small. Will investigate a little bit more.